nuxeo-ai-model/src/main/java/org/nuxeo/ai/model/analyzis/DatasetStatsService.java

```diff
@@ -0,0 +1,119 @@
+
+/**
+ * For a given dataset provides statistics.
+ */
+public interface DatasetStatsService {
+
+    List<String> VOCABULARY_TYPES = Arrays.asList("vocabulary", "xvocabulary", "l10nxvocabulary");
+
+    String AGG_TOTAL = "total";
+
+    Collection<Statistic> getStatistics(CoreSession session, String nxql,
+                                        Set<PropertyType> inputProperties, Set<PropertyType> outputProperties);
+
+    /**
+     * Transforms statistics between ES and UI needs
+     * @param statistics {@link Collection} of {@link Statistic} to transofrm
+     * @return {@link Set} of `unique` {@link FieldStatistics}
+     */
+    default Set<FieldStatistics> transform(Collection<Statistic> statistics) {
+        if (statistics.isEmpty()) {
+            return Collections.emptySet();
+        }
+
+        Map<String, List<Statistic>> collect = statistics.stream().collect(Collectors.groupingBy(Statistic::getField));
+        List<Statistic> totalStat = collect.remove(AGG_TOTAL);
+        collect.remove(AGG_COUNT);
+
+        Set<FieldStatistics> result = new HashSet<>(collect.size());
+        for (Statistic statistic : totalStat) {
+            Number total = statistic.getNumericValue();
+
+            Collection<FieldStatistics> values = collect.entrySet()
+                    .stream()
+                    .flatMap(entry -> entry.getValue().stream())
+                    .collect(Collectors.toMap(Statistic::getField, stat -> FieldStatistics.from(stat, total.longValue()),
+                            FieldStatistics::merge)) // merge all stats to include different aggregates terms|missing|cardinality
+                    .values();
+            result.addAll(values);
+        }
+
+        return result;
+    }
+
+    /**
+     * Checks if given type is a vocabulary.
+     * @param type {@link Type} to retrieve information
+     * @return {@link Boolean#TRUE} if type is a vocabulary
+     */
+    static boolean isVocabulary(Type type) {
+        if (type instanceof ListType) {
+            type = ((ListType) type).getFieldType();
+        }
+        ObjectResolver resolver = type.getObjectResolver();
+        if (resolver instanceof DirectoryEntryResolver) {
+            DirectoryEntryResolver directoryEntryResolver = (DirectoryEntryResolver) resolver;
+            return VOCABULARY_TYPES.contains(directoryEntryResolver.getDirectory().getSchema());
+        }
+        return false;
+    }
+
+    /**
+     * Resolve type of
+     * @param field {@link Field}
+     * @return {@link String} representing one of the field types img|txt|cat
+     */
+    static String getInputType(Field field) {
+        String type;
+        if (field != null && isVocabulary(field.getType())) {
+            type = CATEGORY_TYPE;
+        } else if (field != null && TypeConstants.isContentType(field.getType())) {
+            type = IMAGE_TYPE;
+        } else {
+            type = TEXT_TYPE;
+        }
+        return type;
+    }
+}
```

Comment:
This is the stats part that will need to improve. assuming text or image based on no real info.
But, I guess that to general stats, all that we need to know is if it is blob or not

Reply:
it is a field definition that can be changed during model creation.